Initial SparseIterator experiment

jamwaffles opened the pull request:
Hi! Thank you for helping out with Embedded Graphics development! Please:

- [ ] Check that you've added passing tests and documentation
- [ ] Add a simulator example(s) where applicable
- [ ] Add a `CHANGELOG.md` entry in the **Unreleased** section under the appropriate heading (**Added**, **Fixed**, etc) and appropriate crate (`embedded-graphics`, `simulator`, `tinytga`, `tinybmp`) if your changes affect the **public API**
- [ ] Run `rustfmt` on the project
- [ ] Run `./build.sh` (Linux/macOS only) and make sure it passes. If you use Windows, check that CI passes once you've opened the PR.

## PR description

Experiments with adding a `SparseIterator` struct which wraps a non-contiguous iterator into something that produces an `Option<C>` for each point in a rectangular area. This is the first idea that came to mind and it might be completely the wrong approach, but either way it would be great to get some feedback.

Closes #330

rfuest commented:
> Experiments with adding a `SparseIterator` struct which wraps a non-contiguous iterator into something that produces an `Option<C>` for each point in a rectangular area. This is the first idea that came to mind and it might be completely the wrong approach, but either way it would be great to get some feedback.

I was thinking that it should be implemented the other way around. Take the styled circle iterator for example:
```rust
fn next(&mut self) -> Option<Self::Item> {
    for (point, distance) in &mut self.iter {
        let color = if distance < self.inner_threshold {
            self.inner_color
        } else if distance < self.outer_threshold {
            self.outer_color
        } else {
            None
        };

        if let Some(color) = color {
            return Some(Pixel(point, color));
        }
    }

    None
}
```
Only using the inner part of the loop is a sparse iterator:
```rust
fn next(&mut self) -> Option<Self::Item> {
    self.iter.map(|(point, distance)| {
        if distance < self.inner_threshold {
            self.inner_color
        } else if distance < self.outer_threshold {
            self.outer_color
        } else {
            None
        }
    })
}
```
And it would be wasteful to convert a spare iterator in a non-contiguous iterator just to convert it back again.

jamwaffles commented:
That makes a lot more sense. I've changed `SparseIterator` around and converted the circle rendering to use `fill_sparse` as a first test. I've left the styled circle iterator to continue returning sparse `Pixel`s as a deliberate choice. Do you think we should keep the styled iterator behaviour the same as master (i.e. sparse if necessary) and only add non-sparse code to support `fill_sparse` internally?

rfuest commented:
> I've left the styled circle iterator to continue returning sparse `Pixel`s as a deliberate choice. Do you think we should keep the styled iterator behaviour the same as master (i.e. sparse if necessary) and only add non-sparse code to support `fill_sparse` internally?

The name `StyledPixels` implies that it returns pixels, so if we decide to keep the name it also should return pixels IMO.

The `IntoIterator` trait is very generic and the name `into_iter` doesn't provide any information about what iterator is created. Can we replace it by custom traits to support different iterator types, like we did for the `.points()` method? `styled_primitive.pixels()` is much easier to understand than `styled_primitives.into_iter()` IMO. And we could also add additional methods to create the `Item = Option<C>` iterator variant.

rfuest commented on embedded-graphics/src/draw_target.rs:
```diff
@@ -297,23 +297,22 @@ pub trait DrawTarget {
+    where
+        I: IntoIterator<Item = Option<Self::Color>>,
+    {
+        self.draw_iter(
+            area.points()
+                .zip(colors)
+                .filter_map(|(pos, color)| color.map(|c| Pixel(pos, c))),
+        )
```
Can't this be replaced by using `SparseIterator`?

jamwaffles commented:
The original code was getting pretty stale, so I reset this branch and reimplemented sparse/non-sparse for `Circle`. Can you take a quick look to see if I'm on the right track again?

`FilteredIterator` could _really_ use a better name. It takes a sparse iterator and converts it to a "normal" iterator of `Pixel<C>`. It should be resuable across all primitives, reducing a bit of boilerplate.

jamwaffles commented:
It just occurred to me that we don't necessarily need `FilteredIterator` for our purposes, but it might be useful for others if they want all styled pixels in a primitive without having to filter it themselves. Should we keep it in, along with the `IntoPixels` impl for each primitive?

rfuest commented:
> Can you take a quick look to see if I'm on the right track again?

It's the track we discussed, but maybe not the right one.

I think that `Circle` (and `Ellipse`) could be implemented more efficiently by using a scanline approach using `fill_solid`. This could leverage the symmetry inherit in these shapes to reduce the number of required calculations. And it wouldn't require any caching on the display driver side.

This would only leave text rendering which uses this feature. Perhaps we should just remove the sparse iter entirely and use `draw_iter` instead. The display driver could still try to combine multiple pixels into larger blocks and I think this might not even be much more efficient when using a sparse iterator. With a sparse iter we discard the position information that the display driver needs to recreate and this isn't required when using `draw_iter`. The algorithm for creating larger blocks might be a little more complicated without knowing the iteration order, but this would only needed to be implemented once (in e-g or another crate) and can be shared between multiple display drivers.

jamwaffles commented:
> I think that Circle (and Ellipse) could be implemented more efficiently

Certainly, although for this PR, that code was a precursor to implementing fill_sparse for any other primitives that needed it. A scanline iterator would work a lot better though. It could even use `fill_rect` for each line for a tidy speed boost.

> Perhaps we should just remove the sparse iter entirely and use draw_iter instead.

Ok, so if I understand correctly (sorry, it's very warm here...), you're suggesting closing this PR and forgetting the idea of `fill_sparse` entirely? This would also allow closure of #330 which I think is only missing this `fill_sparse` behaviour.

rfuest commented:
> It could even use `fill_rect` for each line for a tidy speed boost.

I think `fill_rect` is called `fill_solid`. But maybe I'm just imagining things because we have similar temperatures.

> Ok, so if I understand correctly (sorry, it's very warm here...), you're suggesting closing this PR and forgetting the idea of `fill_sparse` entirely? This would also allow closure of #330 which I think is only missing this `fill_sparse` behaviour.

Yes, that's what I'm suggesting. If we change our mind later it wouldn't even be a breaking change because we would provide a default implementation.

jamwaffles commented:
> I think fill_rect is called fill_solid.

Oops yes it is, that's what I meant.

> Yes, that's what I'm suggesting. If we change our mind later it wouldn't even be a breaking change because we would provide a default implementation.

Sweet. I'll close this issue and yeah I agree - adding stuff in later is a lot easier and less breaky than doing it poorly now.

rfuest commented:
@jamwaffles We still need to remove the commented out code in `DrawTarget`. Will you create a PR for that?

jamwaffles commented:
Yep, on it now.
